rfcs/1440_generic_planner_node.md

```diff
@@ -0,0 +1,238 @@
+what the graph can look like by inspecting the code, but it becomes increasingly
+difficult as the complexity of the logic of the planner node increases to cope
+with more complex node graphs. Another drawback of the code-based dynamic node
+creation is that we have to run the code to use the introspection mechanism to
+visualise the node graph.
+
+Generating the node graph automatically based on a configuration/manifest file
+should simplify this. It should be easier to figure out what the node graph will
+look like through inspecting a manifest file (assuming it has an easily
+understandable format). Additional benefits of configuration-based node creation
+include:
+
+- It would be easier to build tools to visualise the node graph without running
+  the application.
+- It would be easier to build tools to perform static analysis of the node graph
+  to detect potential problems.
+- It would usually be easier to reason about the security of a static node
+  graph.
+
+## Overview
+
+Moving all node and channel creation to a configuration-based mechanism is a
+significant and potentially disruptive change. This RFC proposes a more gradual
+approach by implementing a re-usable Wasm-based planner node that uses a
+manifest file to do the node and channel creation. The functionality can be
+expanded over time to support more complex scenarios.
+
+The planner node will be responsible for creating nodes, creating channels,
+sending channel handles to the created nodes to ensure that the full node graph
+is created on start-up. It will will also act as a message router, which would
+route HTTP- and GRPC invocations to other nodes based on the associated labels.
+
+## Detailed Design
+
+### Planner Node
+
+The planner node will the the initial node created for an application. The
+manifest file will be passed to the planner node as part of the application's
+`ConfigMap`.
+
+The planner node will create all the nodes and channels as specified with the
+appropriate labels. It will send initialisation messages to each of the nodes,
+providing additional configuration and handles to channels. It is the
+responsibility of each of the nodes to store these channels for future
+communications with other nodes. It will send a filtered version of the config
+map as part of the initialisation message, based on the configuration keys
+specified in the `config_files` field.
+
+### Invocation Routing
+
+The planner node will also be responsible for routing invocations from the gRPC-
+and HTTP server pseudo nodes to the right receiver node based on the associated
+labels. Since it would be possible for multiple nodes to have the same label,
+the planner node would need an indication of which node should be the invocation
+recipient for a specific label.
+
+### Manifest File
+
+The manifest file will be will be passed to the application as part of the
+initial configuration map as a TOML file. The file will be deserialised into the
+following structure:
+
+```rust
+
```

Comment:
Perhaps briefly explain how this interacts with [defenes-in-depth](https://github.com/project-oak/oak/issues/1029) (which is still work in progress). I think some of the flags in `Node` may interact with some of the flags in `PermissionsConfiguration`:
https://github.com/project-oak/oak/blob/bf82a0947bb1eb7ef4f5f6133c25e2e2837b8114/oak_runtime/src/permissions.rs#L26

This might actually be more relevant for the analyzer tools listed below.

Reply:
I added some mention here and in the analysis tools section.